packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,297 @@
+    const emitChange = debounce(() => {
+      this.emit('change', Array.from(changedPages))
+      changedPages.clear()
+    }, 100)
+
+    const helpers: PageHelpers = {
+      extractStaticData,
+      loadPageData: (file) => Promise.resolve(loadPageData(file, helpers)),
+      addPageData,
+      watchFiles,
+    }
+
+    // The file currently being processed.
+    let currentFile: File | null = null
+
+    let pagesPromise = Promise.resolve(pageCache)
+    findPages(pagesDir, helpers)
+
+    this.getPages = () => pagesPromise
+    this.close = () => watchers.forEach((w) => w.close())
+
+    function addPageData({
+      pageId,
+      key = 'main',
+      dataPath,
+      staticData,
+    }: PageData) {
+      if (!pageId.startsWith('/'))
+        throw Error(
+          `addPageData error: pageId should start with "/", but got "${pageId}"`
+        )
+
+      const existingPage = pageCache[pageId]
+      const page =
+        existingPage ||
+        (pageCache[pageId] = {
+          data: {},
+          staticData: {},
+        })
+
+      if (dataPath) {
+        if (page.data[key])
+          throw Error(
+            `addPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = dataPath
+      }
+
+      if (staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `addPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = staticData
+      }
+
+      if (currentFile) {
+        let dataKeys = currentFile.dataKeys.get(pageId)
+        if (!dataKeys) currentFile.dataKeys.set(pageId, (dataKeys = []))
+        dataKeys.push(key)
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    function removePageData(dataKeys: string[], pageId: string) {
+      const page = pageCache[pageId]
+
+      dataKeys.forEach((key) => {
+        delete page.data[key]
+        delete page.staticData[key]
+      })
+
+      if (hasKeys(page.data) || hasKeys(page.staticData)) {
+        changedPages.add(pageId)
+        emitChange()
+      } else {
+        delete pageCache[pageId]
+      }
+    }
+
+    function hasKeys(obj: object) {
+      for (const _ in obj) return true
+      return false
+    }
+
+    function watchFiles(
+      baseDir: string,
+      arg2?: string | string[] | FileHandler,
+      arg3?: FileHandler
+    ) {
+      let globs: string[]
+      let handler: FileHandler
+
+      if (typeof arg2 === 'function') {
+        globs = ['**/*']
+        handler = arg2
+      } else {
+        globs = Array.isArray(arg2) ? arg2 : [arg2 || '**/*']
+        handler = arg3 || defaultPageLoader
+      }
+
+      // Strip trailing slash and make absolute
+      baseDir = path.resolve(pagesDir, baseDir)
+
+      const processFile = async (
+        type: 'add' | 'change' | 'unlink',
+        file: File
+      ) => {
+        file.dataKeys.forEach(removePageData)
```

Comment:
Why do we do this for all change type?

Reply:
It's a no-op for `add` events, since `dataKeys` will be empty